### PR TITLE
Chapter2

### PR DESCRIPTION
**Pull request by godu**
### Solutions for Chapter 2

cc @vrom911 @chshersh

### CODE REVIEW

File: src/Chapter2.hs [hunk 1]
```diff
@@ -858,7 +875,9 @@ and reverses it.
   function, but in this task, you need to implement it manually. No
   cheating!
 -}
-rewind = error "rewind: Not Implemented!"
+rewind :: [a] -> [a]
+rewind [] = []
+rewind (x:xs) = rewind xs ++ [x]
```
Comment (vrom911) [inline src/Chapter2.hs:880]: Your solution is correct! However, it is slow. In lists, it is quite slow to add anything at the end of the list. That is why it is always better to rewrite it with the `:` cons. Remember the explanation with trains? 🚂 🚋 🚋 

That is why a more efficient solution is with the accumulator and the recursive function that will do the addition at the start of the list which is instant!

You can read a bit more about the `go` pattern in here: https://kowainik.github.io/posts/haskell-mini-patterns#recursive-go

File: src/Chapter2.hs [hunk 2]
```diff
@@ -842,7 +857,9 @@ list.
 
 🕯 HINT: Use the 'cycle' function
 -}
-rotate = error "rotate: Not implemented!"
+rotate :: Int -> [a] -> [a]
+rotate x _ | x < 0 = []
```
Comment (vrom911) [inline src/Chapter2.hs:861]: Guards are better when used with several cases, that would make it much more readable, also you can reuse the variables.

File: src/Chapter2.hs [hunk 3]
```diff
@@ -842,7 +857,9 @@ list.
...
+rotate x xs = take (length xs) $ drop x $ cycle xs
```
Comment (vrom911) [inline src/Chapter2.hs:862]: You could have a bit more efficient solution if you would check on an empty list and if `x` is zero (by module of the length as well) 🙂

**Review by vrom911 — APPROVED** [review]
Nice Chapter! I added a few suggestions to make your solutions faster and more canonical 🙂 👍🏼

File: src/Chapter2.hs [hunk 4]
```diff
@@ -842,7 +854,10 @@ list.
 
 🕯 HINT: Use the 'cycle' function
 -}
-rotate = error "rotate: Not implemented!"
+rotate :: Int -> [a] -> [a]
+rotate x l | x < 0 = []
+           | null l = []
+           | otherwise = take (length xs) $ drop x $ cycle xs
```
Comment (vrom911) [inline src/Chapter2.hs:860]: I guess, it is `l` in here instead of `xs` 🙂

**Comment by chshersh** [comment]
Thanks for participating in Learn4Haskell! Hope you enjoyed this Hacktoberfest 🥳 

I am going to close this PR as we can't merge the solutions into our course 🔐 

Even though the Hacktoberfest is over, we are going to continue reviewing, so feel free to open new PRs, if you haven't finished the course yet! 🎊

**Pull request closed without merging**